### PR TITLE
feat: Add genesis support for new modules and group genesis

### PR DESCRIPTION
**Pull request by blushi**
closes: #214

### CODE REVIEW

File: types/module/server/module.go
Comment (aaronc) [inline types/module/server/module.go:23]:
```suggestion
	RegisterGenesisHandlers(module.InitGenesisHandler, module.ExportGenesisHandler)
```

File: testutil/fixture.go
Comment (aaronc) [inline testutil/fixture.go:53]: Why do we need all this here?

Comment (blushi) [reply testutil/fixture.go:53]: To use them in tests, see in `x/group/server/testsuite/` but happy to discuss any other solutions that would look better to you.

Comment (aaronc) [reply testutil/fixture.go:53]: How about we simply have `InitGenesis()` and `ExportGenesis()` methods? Rather than being able to look up the handler by module. Also consider we want to use these same `Fixture`s for `Network` based tests in the future (#246).

Comment (blushi) [reply testutil/fixture.go:53]: The thing is that we can register multiple modules as part of `FixtureFactory` `Setup()` which is why I added them as look up methods.

**Comment by aaronc** [comment]
I'm planning to review this soon. In the meantime can anyone else review?

**Comment by blushi** [comment]
> I'm planning to review this soon. In the meantime can anyone else review?

@amaurym said he'll have a look.

File: testutil/fixture.go
Comment (aaronc) [reply testutil/fixture.go:53]: I don't think that approach will work using a `Network` fixture. Would `InitGenesis` and `ExportGenesis` also work for the tests?

**Review by amaurym — COMMENTED** [review]
LGTM overall. Would like to avoid panics, and after that for me we can merge.

I also added a bunch of questions, mostly me getting up to speed with the groups module.

File: x/group/server/testsuite/genesis.go
Comment (blushi) [reply x/group/server/testsuite/genesis.go:154]: No this won't always be the same values, for example you could have one master group and then multiple group accounts for that group with different admins, metadata and versions.

File: x/group/server/testsuite/genesis.go
Comment (blushi) [reply x/group/server/testsuite/genesis.go:119]: I'll update so that we only have `s.exportGenesisHandler(ctx, cdc)` as part of it

File: proto/regen/group/v1alpha1/genesis.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/genesis.proto:14]: Let's say that in the future we want to be able to delete groups then we won't necessarily be able to deduct `group_seq` from `len(groups)` so I'd rather keep it.

File: x/group/server/genesis.go
Comment (blushi) [reply x/group/server/genesis.go:19]: the `groupTable` doesn't implement `SequenceExportable` (see `orm/genesis.go`) and the seq value is only usable in this case, which is why we initialize it after using `s.groupSeq.InitVal`directly.

File: x/group/server/server.go
Comment (blushi) [reply x/group/server/server.go:168]: good point, I'll try to improve that, thanks!

File: x/group/server/testsuite/genesis.go
Comment (blushi) [reply x/group/server/testsuite/genesis.go:1]: I follow the existing structure and I just didn't want to put the genesis test in the already huge `suite.go` file. I don't mind adding `_test` suffix but then if you try to run the tests individually in `/testsuite` folder, it won't find any tests to run because the actual test with `suite.Run` is in `server_test.go` which is in a different package.

File: x/group/server/server.go
Comment (blushi) [reply x/group/server/server.go:168]: Though we can't really add `RegisterMsgServer` and `RegisterQueryServer` as part of the interface, these are proto-generated standalone functions, `serverImpl` rather implements `MsgServer` and `QueryServer` interfaces, which are module specific. So I'd prefer to have separate interfaces for different concerns, and create a separate interface for genesis which would look like this (not really sure about the naming here):
````go
type GenesisServer interface {
  InitGenesis() ...
  ExportGenesis() ...
}
```

Comment (blushi) [reply x/group/server/server.go:168]: Or we could also have a single interface as you suggested such as:
```go
type Server interface {
 InitGenesis() 
 ExportGenesis() 
 RegisterInvariants()
 ...
}
```
And then we could also have one single entry method for the `configurator` to register genesis handlers, invariants, etc. 
```go
func (c *configurator) RegisterGenesisHandlers(s Server) {
 c.initGenesisHandler = s.InitGenesis
 c.exportGenesisHandler = s.ExportGenesis
 c.registerInvariantsHandler = s.RegisterInvariants
 ...
}
```

But maybe that can be done as part of a separate PR, once we have more functions to register, wdyt?

File: testutil/fixture.go
Comment (blushi) [reply testutil/fixture.go:53]: When you say `InitGenesis()` and `ExportGenesis()` methods, could you elaborate on the params and return values? 
Do you see them as being similar to the manager  `InitGenesis()` and `ExportGenesis()`, ie they would actually execute  init or export genesis (instead of simply returning a function)?

File: x/group/server/testsuite/genesis.go
Comment (amaurym) [reply x/group/server/testsuite/genesis.go:1]: Ah true. Hmm, okay let's keep it like this for now.

File: x/group/server/server.go
Comment (amaurym) [reply x/group/server/server.go:168]: > But maybe that can be done as part of a separate PR, once we have more functions to register, wdyt?

Yes, good idea. https://github.com/regen-network/regen-ledger/issues/301

**Review by amaurym — APPROVED** [review]
All my questions were addressed, thanks @blushi

File: x/group/server/server.go
Comment (robert-zaremba) [reply x/group/server/server.go:168]: Do we need this things at all in a `configurator`? In comment's above I'm trying to find out if we could reuse `module.Manager`.

File: x/group/server/testsuite/genesis.go
Comment (robert-zaremba) [inline x/group/server/testsuite/genesis.go:171]: in new tests, how about using `
```
require := s.Require()
...
require.Equal(...)
```
It makes the code easier to read (less parenthesis) and a bit shorter .

File: proto/regen/group/v1alpha1/genesis.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/genesis.proto:17]: info is uncountable

Comment (blushi) [reply proto/regen/group/v1alpha1/genesis.proto:17]: but maybe we could update it to this instead?
```suggestion
  // groups is the list of groups info.
```

File: proto/regen/group/v1alpha1/genesis.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/genesis.proto:27]: ditto

File: app/app.go
Comment (blushi) [reply app/app.go:197]: it was for new manager, but I actually prefer `smm` (`sm` is already used for simulation)

File: app/stable_appconfig.go
Comment (blushi) [reply app/stable_appconfig.go:17]: it's actually setting something in experimental mode (see `app/experimental_appconfig.go`)

File: orm/auto_uint64_test.go
Comment (blushi) [reply orm/auto_uint64_test.go:1]: Moved them to `orm_test.go` to avoid import cycles.

File: orm/iterator_test.go
Comment (blushi) [reply orm/iterator_test.go:1]: https://github.com/regen-network/regen-ledger/pull/270#discussion_r600409421

File: proto/regen/group/v1alpha1/types.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: groups and group accounts are not the same thing, but we rename it to `group_account_address`,  I agree that would be clearer.

Comment (robert-zaremba) [reply proto/regen/group/v1alpha1/types.proto:137]: Yes, let's rename. We are already in the `Group` module context. Do we need to put `account`? How about `group_address` or `g_address`? It's documented, so the doc string will stay in the structure.

Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: we can have several group accounts for the same group so I would find it confusing to just name it `group_address`

File: app/stable_appconfig.go
Comment (robert-zaremba) [reply app/stable_appconfig.go:17]: ah, we have a build flag. Then please add a doc string (to both functions), it's not very clear if we go to this code and expect to see if there is same function in other build configuration.

File: proto/regen/group/v1alpha1/types.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: since we're already in the `GroupAccount` definition, maybe just `address` would be fine?

Comment (robert-zaremba) [reply proto/regen/group/v1alpha1/types.proto:137]: Sounds good. doc comment should be updated as well.

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go:41]: > * use `[]NaturalKeyed`  (that would require copying slice all whenever we call this function - which shouldn't be a big deal, because we will only copy `len(slice)` pointers).

As you said, for example that would mean that we would need to copy the slice of `GroupInfo` into a slice of `Model` before calling the `ImportTableData` function, and I don't find this really good from a dev UX point of view. Also it would just clutter the `InitGenesis` function in `x/group/server/genesis.go`, because we would need to add that for all the table imports.

> * or an interator pattern:
> 
> ```
> type NaturalKeyedIterator interface {
>    Next() bool   // goes to the next element, returns false if there are no more elements
>    Get() NaturalKeyed // returns current element
> }
> 
> type MyObjs {
>    slice []MyObj
>    idx uint
> }
> func (m *MyObjs) Next () { 
>     if m.idx >= len(slice) - 1 { return false }
>     m.idx++
>     return true
> } 
> func (m *MyObjs) Get() NaturalKeyed {  return m.slice[m.idx] }
> ```

I prefer this second approach, but that would mean that we would need to define a struct such as `MyObjs` for all models, eg Groups, GroupAccounts, etc. while their `NaturalKeyedIterator` implementation logic would all be the same.

File: proto/regen/group/v1alpha1/types.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: We also use `group_account` in `Proposal` and `MsgCreateProposalRequest` so I think it would make sense in those cases to keep having some reference to the group account in the naming, whether we use `group_account_address` or just `account_address`, wdyt?

Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: Or maybe we just use `address` everywhere to keep things consistent, since there's only one address that you can provide in these cases, which is the group account address...

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go:43]: We're not only importing rows but also the table sequence when applicable so I'd rather keep the current naming.

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go]: @robert-zaremba I have changed it to "table PrefixScan failure when exporting table data", wdyt?

File: testutil/testdata/model.go
Comment (blushi) [reply testutil/testdata/model.go]: makes sense.

File: types/module/server/manager.go
Comment (blushi) [reply types/module/server/manager.go:26]: the manager `InitGenesis` and `ExportGenesis` methods use those handlers.

File: types/module/server/manager.go
Comment (blushi) [reply types/module/server/manager.go:175]: This server manager is used for the new modules (ecocredit, data, group) which are not registered with the usual sdk manager (see module wiring in `app.go` and `experimental_appconfig.go`). These new modules don't implement `AppModule` and don't have a direct access to the server implementation which is private (contrary to sdk modules with a reference to a keeper). So we pass relevant methods (init/export genesis in this case) to the `configurator` on `RegisterServices` so they can be passed in turn to the server manager on `RegisterModules`, then the server manager can use them to perform init/export genesis when needed.

File: x/group/server/server.go
Comment (blushi) [reply x/group/server/server.go:168]: see https://github.com/regen-network/regen-ledger/pull/270/#discussion_r601267369 and sample code: https://github.com/regen-network/regen-ledger/blob/789f363747892d0436df09cba863b29dbc00d8d4/types/module/server/manager.go#L106-L108

File: testutil/fixture.go
Comment (blushi) [reply testutil/fixture.go]: I've updated it so we have `InitGenesis` and `ExportGenesis` methods instead.

File: orm/genesis.go
Comment (robert-zaremba) [reply orm/genesis.go:41]: I think copying indexes (first approach) is not bad. It won't take much more memory since we are only copying pointers.

If you like we can do it (or consider 2nd approach or even something else) later in a new issue?

File: types/module/server/manager.go
Comment (robert-zaremba) [reply types/module/server/manager.go:186]: Could you add a doc string explaining that (to the `Manager.Init/Export Genesis`

Comment (robert-zaremba) [reply types/module/server/manager.go:186]: Also, this mechanism looks tricky. Instead of calling interface, we pass a manager and ask the module to register itself. I don't think that extending the Registration pattern is a good approach. Not going to block on that. Let me know wht do you think. If we find that it's good to rethink it then maybe we can continue in a new issue?
@aaronc  - do you have any comments on that?

**Review by robert-zaremba — COMMENTED** [review]
Let's think what to do with `types/module.Manager` vs `initGenesisHandlers` and `exportGenesisHandlers`
Not going to block on it, but let's see if this is a right approach we want to use towards SDKv1 before merging. (we can expand on this in an issue).

File: app/stable_appconfig.go
Comment (blushi) [reply app/stable_appconfig.go:17]: added

File: types/module/server/manager.go
Comment (aaronc) [reply types/module/server/manager.go:186]: Maybe we should try something like I'm describing here which is a hybrid between the ADR 033 approach and the existing approach: https://github.com/regen-network/regen-ledger/issues/301#issuecomment-804179925.

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go:41]: Ok I'll create an issue

File: proto/regen/group/v1alpha1/types.proto
Comment (blushi) [reply proto/regen/group/v1alpha1/types.proto:137]: @robert-zaremba up

File: types/module/server/manager.go
Comment (blushi) [reply types/module/server/manager.go:186]: Let's follow up on that as part of this issue https://github.com/regen-network/regen-ledger/issues/301

File: proto/regen/group/v1alpha1/types.proto
Comment (robert-zaremba) [reply proto/regen/group/v1alpha1/types.proto:137]: As long we have a right comment for the struct field it will be good. Let's do that.

File: types/module/server/manager.go
Comment (robert-zaremba) [reply types/module/server/manager.go:186]: Yes, let's roll this PR and think about better approach in the issue.

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go:41]: https://github.com/regen-network/regen-ledger/issues/308

Comment (robert-zaremba) [reply orm/genesis.go:41]: Oh, I did it in the meantime. https://github.com/regen-network/regen-ledger/issues/309

**Review by robert-zaremba — APPROVED** [review]
Please recheck if we have documentation for attributes updated.

File: orm/genesis.go
Comment (blushi) [reply orm/genesis.go:41]: ok I'll close mine

**Pull request merged**
